File: mobile/lib/models/daily_challenge.dart
/// DailyChallenge — one daily prediction/scenario question with options
/// and an explanation that surfaces after the user submits.
library;

import 'package:equatable/equatable.dart';

class DailyChallenge extends Equatable {
  const DailyChallenge({
    required this.id,
    required this.type,
    required this.difficulty,
    required this.locale,
    required this.scenario,
    required this.question,
    required this.options,
    required this.answer,
    required this.explanation,
    this.relatedLesson,
    this.relatedAgent,
    this.tags = const [],
  });

  final String id;
  final String type;
  final int difficulty;
  final String locale;
  final String scenario;
  final String question;
  final List<String> options;
  final int answer;
  final String explanation;
  final String? relatedLesson;
  final String? relatedAgent;
  final List<String> tags;

  factory DailyChallenge.fromJson(Map<String, dynamic> j) => DailyChallenge(
        id: j['id'] as String,
        type: j['type'] as String,
        difficulty: j['difficulty'] as int,
        locale: j['locale'] as String? ?? 'en',
        scenario: j['scenario'] as String,
        question: j['question'] as String,
        options: List<String>.from(j['options'] as List),
        answer: j['answer'] as int,
        explanation: j['explanation'] as String,
        relatedLesson: j['related_lesson'] as String?,
        relatedAgent: j['related_agent'] as String?,
        tags: List<String>.from((j['tags'] as List?) ?? const []),
      );

  @override
  List<Object?> get props => [id];
}


/// CR010 (B5) — the caller's stored attempt for today's challenge, from
/// `GET /today`'s `my_attempt`. Null when unauthenticated or unanswered.
class MyAttempt extends Equatable {
  const MyAttempt({
    required this.selectedOption,
    required this.correct,
    required this.attemptedAt,
  });

  final int selectedOption;
  final bool correct;
  final String attemptedAt; // ISO 8601

  factory MyAttempt.fromJson(Map<String, dynamic> j) => MyAttempt(
        selectedOption: (j['selected_option'] as num).toInt(),
        correct: j['correct'] as bool,
        attemptedAt: j['attempted_at'] as String? ?? '',
      );

  @override
  List<Object?> get props => [selectedOption, correct, attemptedAt];
}


class DailyChallengeWithDate extends Equatable {
  const DailyChallengeWithDate({
    required this.challenge,
    required this.date,
    this.myAttempt,
  });

  final DailyChallenge challenge;
  final String date; // YYYY-MM-DD
  final MyAttempt? myAttempt; // CR010 (B5): server-truth prior attempt

  factory DailyChallengeWithDate.fromJson(Map<String, dynamic> j) =>
      DailyChallengeWithDate(
        challenge:
            DailyChallenge.fromJson(j['challenge'] as Map<String, dynamic>),
        date: j['date'] as String,
        myAttempt: j['my_attempt'] == null
            ? null
            : MyAttempt.fromJson(j['my_attempt'] as Map<String, dynamic>),
      );

  @override
  List<Object?> get props => [challenge.id, date, myAttempt];
}


/// `POST /v1/daily_challenge/{id}/attempt` response — server-graded result.
/// `alreadyAttempted` is true when the challenge was answered before; the
/// result then reflects the STORED attempt, not a re-grade of the new answer.
class DailyChallengeAttemptResult extends Equatable {
  const DailyChallengeAttemptResult({
    required this.correct,
    required this.correctOption,
    required this.explanation,
    required this.selectedOption,
    required this.alreadyAttempted,
    this.relatedLesson,
    this.relatedAgent,
  });

  final bool correct;
  final int correctOption;
  final String explanation;
  final int selectedOption;
  final bool alreadyAttempted;
  final String? relatedLesson;
  final String? relatedAgent;

  factory DailyChallengeAttemptResult.fromJson(Map<String, dynamic> j) =>
      DailyChallengeAttemptResult(
        correct: j['correct'] as bool? ?? false,
        correctOption: (j['correct_option'] as num?)?.toInt() ?? 0,
        explanation: j['explanation'] as String? ?? '',
        selectedOption: (j['selected_option'] as num?)?.toInt() ?? 0,
        alreadyAttempted: j['already_attempted'] as bool? ?? false,
        relatedLesson: j['related_lesson'] as String?,
        relatedAgent: j['related_agent'] as String?,
      );

  @override
  List<Object?> get props =>
      [correct, correctOption, selectedOption, alreadyAttempted];
}

File: backend/app/schemas/daily_challenge.py
"""Daily-challenge schemas.

One JSON file per month at `content/daily_challenges/YYYY_MM.json` — a flat
list of challenge objects. The id format is `dc_YYYY_MM_DD_<slug>`, so the
day number is encoded in the id. Server-side "today" selection reads the
canonical Asia/Kuala_Lumpur date and matches the prefix.

Each challenge is a single-question multiple-choice prompt with one correct
answer index, plus an explanation that surfaces after the user submits.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class DailyChallenge(BaseModel):
    id: str
    type: str
    difficulty: int = Field(ge=1, le=5)
    locale: str = "en"
    scenario: str
    question: str
    options: list[str]
    answer: int = Field(ge=0)
    explanation: str
    related_lesson: str | None = None
    related_agent: str | None = None
    tags: list[str] = Field(default_factory=list)


class DailyChallengePublic(BaseModel):
    """The pre-attempt public shape of a challenge — `DailyChallenge` minus
    `answer` AND `explanation` (DEF042: `/today` returned both, so an
    API-direct user could either read `answer` outright or match the
    always-correct-restating `explanation` text against the options —
    either way the correct choice was derivable before attempting, which
    is farmable for the `challenge_correct` streak/credit milestone).
    Server-side grading still uses the full `DailyChallenge`; only the
    response-facing views built from this class drop the giveaway fields.
    """

    id: str
    type: str
    difficulty: int = Field(ge=1, le=5)
    locale: str = "en"
    scenario: str
    question: str
    options: list[str]
    related_lesson: str | None = None
    related_agent: str | None = None
    tags: list[str] = Field(default_factory=list)

    @classmethod
    def from_challenge(cls, ch: DailyChallenge) -> "DailyChallengePublic":
        return cls(**ch.model_dump(exclude={"answer", "explanation"}))


class MyAttempt(BaseModel):
    """The caller's stored answer for a challenge (CR004 — server truth;
    lets the mobile card render the answered state after a restart).
    Carries `correct_option` + `explanation` too (DEF042) — safe here
    because this only ever populates for a challenge the caller has
    already attempted."""

    selected_option: int
    correct: bool
    correct_option: int
    explanation: str
    attempted_at: str  # ISO 8601


class DailyChallengeResponse(BaseModel):
    challenge: DailyChallengePublic
    date: str  # YYYY-MM-DD
    my_attempt: MyAttempt | None = None


class DailyChallengeListResponse(BaseModel):
    items: list[DailyChallengePublic]
    total: int
